Framework/DataHandling/src/LoadSampleShape.cpp

```diff
@@ -0,0 +1,222 @@
+  for (uint16_t i = 0; i < vertices.size(); ++i) {
+    if (areEqualVertices(vertex, vertices[i])) {
+      return i;
+    }
+  }
+  vertices.push_back(vertex);
+  uint16_t index = static_cast<uint16_t>(vertices.size() - 1);
+  if (index != vertices.size() - 1) {
+    throw std::runtime_error("Too many vertices in solid");
+  }
+  return index;
+}
+
+std::unique_ptr<MeshObject> readSTLMeshObject(std::ifstream &file) {
+  std::vector<uint16_t> triangleIndices;
+  std::vector<V3D> vertices;
+  V3D t1, t2, t3;
+
+  while (readSTLTriangle(file, t1, t2, t3)) {
+    // Add triangle if all 3 vertices are distinct
+    if (!areEqualVertices(t1, t2) && !areEqualVertices(t1, t3) &&
+        !areEqualVertices(t2, t3)) {
+      triangleIndices.push_back(addSTLVertex(t1, vertices));
+      triangleIndices.push_back(addSTLVertex(t2, vertices));
+      triangleIndices.push_back(addSTLVertex(t3, vertices));
+    }
+  }
+  // Use efficient constructor of MeshObject
+  std::unique_ptr<MeshObject> retVal = std::unique_ptr<MeshObject>(
+      new MeshObject(std::move(triangleIndices), std::move(vertices),
+                     Mantid::Kernel::Material()));
+  return retVal;
+}
+
+std::unique_ptr<Geometry::MeshObject> readSTLSolid(std::ifstream &file,
+                                                   std::string &name) {
+  // Read Solid name
+  // We expect line after trimming to be "solid "+name.
+  std::string line;
+  if (getline(file, line)) {
+    boost::trim(line);
+    if (line.size() < 5 || line.substr(0, 5) != "solid") {
+      throw std::runtime_error("Expected start of solid");
+    } else {
+      name = line.substr(6, std::string::npos);
+    }
+    // Read Solid shape
+    return readSTLMeshObject(file);
+  }
+  return nullptr;
+}
+
+} // end anonymous namespace
+
+void LoadSampleShape::init() {
+  auto wsValidator = boost::make_shared<API::InstrumentValidator>();
+  ;
+
+  // input workspace
+  declareProperty(
+      make_unique<WorkspaceProperty<>>("InputWorkspace", "", Direction::Input,
+                                       wsValidator),
+      "The name of the workspace containing the instrument to add the shape");
+
+  // shape file
+  const std::vector<std::string> extensions{".stl"};
+  declareProperty(
+      make_unique<FileProperty>("Filename", "", FileProperty::Load, extensions),
+      "The path name of the file containing the shape.");
+
+  // Output workspace
+  declareProperty(make_unique<WorkspaceProperty<>>("OutputWorkspace", "",
+                                                   Direction::Output),
+                  "The name of the workspace that will contain the loaded "
+                  "shape of the sample");
+}
+
+/**
+* Return the confidence with with this algorithm can load the file
+* @param descriptor A descriptor for the file
+* @returns An integer specifying the confidence level. 0 indicates it will not
+* be used
+*/
+int LoadSampleShape::confidence(Kernel::FileDescriptor &descriptor) const {
```

Comment:
There is a question if you need this at all, but if you do this is a simplistic and only partial implementation.  In the exec, you throw an error if the first part of the first line is not "SOLID", this should be checked here.